src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/MemorySegmentPool.java

```diff
@@ -0,0 +1,375 @@
+    int bound = calculateBucket(alignedSize);
+    final var segmentEntry = findOrAllocateItem(bound);
+
+    ((ResourceScopeImpl) resourceScope).addOrCleanupIfFail(new ResourceCleanup() {
+      @Override
+      @ForceInline
+      public void cleanup() {
+        segmentEntry.close();
+      }
+    });
+
+    return segmentEntry;
+  }
+
+  /**
+   * Calculates bit bound, of size - in other words the bucket which should be used for item.
+   * @param alignedSize
+   */
+  @ForceInline
+  protected int calculateBucket(long alignedSize) {
+    // If 100.., than 100... - 1 -> 01111
+    // If 101 -> than 101 - 1 -> 1....
+
+    // 0 -> 64
+    // This equation does not allow to allocate more than 2^63, however such memory may require
+    // 5 level page cache, so skippable for now
+    return 64 - Long.numberOfLeadingZeros(alignedSize - 1);
+  }
+
+  /**
+   * Allocates and prepares a new item.
+   */
+  @ForceInline
+  private MemoryPoolItem allocateNewPoolItem(SpinLockQueue<MemoryPoolItem> queue, long allocationSize) {
+    final var memoryAddress = CLinker.allocateMemory(allocationSize);
+    return new MemoryPoolItem(queue, memoryAddress, allocationSize, scope);
+  }
+
+  /**
+   * Free all elements associated with pool. Called when pool's scope gets closed.
+   */
+  private void freePool() {
+
+  }
+
+  private static void validateMaxSizes(int maxSizes[]) {
+    Arrays.stream(maxSizes).filter(i -> i < 0).findAny()
+        .ifPresent(i -> {
+          throw new IllegalStateException("Invalid max size " + i);
+        });
+  }
+
+  private static long alignSize(long bytesSize, long bytesAlignment) {
+    return (bytesSize + bytesAlignment - 1) & -bytesAlignment;
+  }
+
+  /**
+   * Represent single item in the pool with related data.
+   * <br />
+   * In order to return item back to pool close should be called (depending on context either
+   * implicite or explicite).
+   */
+  public static final class MemoryPoolItem extends SpinLockQueue.Entry<MemoryPoolItem> implements AutoCloseable {
+    /** The owning queue (bucket) to which this pool item belongs. */
+    private final SpinLockQueue<MemoryPoolItem> owner;
+
+    /** Source memory address, it's the start of sourceSegment. */
+    private final MemoryAddress sourceAddress;
+
+    /** Flag indicating if item has been released. */
+    private volatile boolean released;
+
+    private final static VarHandle RELEASED;
+
+    static {
+      try {
+        RELEASED = MethodHandles.lookup().findVarHandle(MemoryPoolItem.class, "released", boolean.class);
+      } catch (Exception e) {
+        throw new ExceptionInInitializerError(e);
+      }
+    }
+
+    @ForceInline
+    private MemoryPoolItem(SpinLockQueue<MemoryPoolItem> queue, MemoryAddress sourceAddress, long size, ResourceScope scope) {
+      super();
+      this.owner = queue;
+      this.sourceAddress = sourceAddress;
+
+      ((ResourceScopeImpl) scope).addOrCleanupIfFail(new ResourceCleanup() {
```

Comment:
Not 100% sure you need this. You already have added a cleanup action which calls `ClearRunnable`. Why not just deallocating stuff from there? In fact, I think the scope should be also moved to the spin lock queue - each queue has a scope, and when the scope is closed, the queue is locked, and all entries inside it are released. This will save a cleanup action addition on every new memory pool item (that's how I'd do it at least).

Reply:
That's very fair point!!! Thanks.

After calling cleaner queue max size is set to 0, so any pending entry will have to be released on close.

Outstanding entires can be closed using iterator.